packages/ember-metal/lib/properties.js

```diff
@@ -83,61 +83,71 @@ export function DEFAULT_GETTER_FUNCTION(name) {
   @param {*} [data] something other than a descriptor, that will
     become the explicit value of this property.
 */
-export function defineProperty(obj, keyName, desc, data, meta) {
-  var possibleDesc, existingDesc, watching, value;
+export function defineProperty(obj, keyName, desc, data/*, meta*/) {
+  let meta = arguments[4] || metaFor(obj);
+  let watchEntry = meta.peekWatching(keyName);
+  let existingDesc = meta.peekDescs(keyName);
 
-  if (!meta) {
-    meta = metaFor(obj);
-  }
-  var watchEntry = meta.peekWatching(keyName);
-  possibleDesc = obj[keyName];
-  existingDesc = (possibleDesc !== null && typeof possibleDesc === 'object' && possibleDesc.isDescriptor) ? possibleDesc : undefined;
-
-  watching = watchEntry !== undefined && watchEntry > 0;
+  let watching = watchEntry !== undefined && watchEntry > 0;
 
   if (existingDesc) {
     existingDesc.teardown(obj, keyName);
   }
 
+  let value;
+
+  // is this instanceof needed?
   if (desc instanceof Descriptor) {
-    value = desc;
-    if (isEnabled('mandatory-setter')) {
-      if (watching) {
-        Object.defineProperty(obj, keyName, {
-          configurable: true,
-          enumerable: true,
-          writable: true,
-          value: value
-        });
-      } else {
+    meta.writeDescs(keyName, desc);
```

Comment:
we should likely extract this algo into `installDescriptor` function or something